src/visualizers/VisualizerGrid.ts

```diff
@@ -0,0 +1,1220 @@
+        }
+
+        return sumOfDivisors
+    }
+
+    isSumOfTwoSquares(ind: number): boolean {
+        const factors = this.seq.getFactors(ind)
+
+        if (factors === null) {
+            return false // we can't factor, so can't tell
+        }
+        let legendre = 1
+        for (let i = 0; i < factors.length; i++) {
+            const factor = factors[i]
+            const prime = factor[0]
+            const exponent = factor[1]
+            if (prime === -1n) {
+                // negative numbers are never sums of squares
+                return false
+            }
+            // otherwise update legendre for every prime 3 mod 4
+            if (prime % 4n === 3n && exponent % 2n === 1n) {
+                legendre = -1 * legendre
+            }
+        }
+        if (legendre === 1) {
+            return true
+        } else {
+            return false
+        }
+    }
+
+    //Modification of Geeks for Geeks :
+    //https://www.geeksforgeeks.org/program-check-n-pentagonal-number/
+    isPolygonal(ind: number, order: bigint): boolean {
+        const num = this.seq.getElement(ind)
+
+        // negative inputs are never polygonal
+        if (num < 0n) {
+            return false
+        }
+        let i = 1n,
+            M
+        do {
+            M = (order - 2n) * ((i * (i - 1n)) / 2n) + i
+            i += 1n
+        } while (M < num)
+        return M === num
+    }
+
+    isAbundant(ind: number): boolean {
+        const num = this.seq.getElement(ind)
+        return this.getSumOfDivisors(num) - num > num
+    }
+
+    isPerfect(ind: number): boolean {
+        const num = this.seq.getElement(ind)
+        const sum = this.getSumOfDivisors(num) - num
+        // If sum of divisors is equal to
+        // n, then n is a perfect number
+        if (sum === num && num != 1n) {
+            return true
+        }
+
+        return false
+    }
+
+    isSemiPrime(ind: number): boolean {
+        const factors = this.seq.getFactors(ind)
+        if (
+            factors === null // if we can't factor, it isn't semi-prime
+            || factors.length === 0 // 1 is not semi-prime
+            || factors[0][0] === 0n // 0 is not semi-prime
+            || (factors.length === 1 && factors[0][0] === -1n) // -1 not s-prime
+        ) {
+            return false
+        }
+        if (
+            (factors.length === 1 && factors[0][1] === 2n) // square
+            || (factors.length === 2
+                && factors[0][0] === -1n
+                && factors[1][1] === 2n) // negative of square
+            || (factors.length === 2
+                && factors[0][1] === 1n
+                && factors[1][1] === 1n) // semi-prime
+            || (factors.length === 3
+                && factors[0][0] === -1n
+                && factors[1][1] == 1n
+                && factors[2][1] == 1n) // negative of semi-prime
+        ) {
+            return true
+        } else {
+            return false
+        }
+    }
+}
+
```

Comment:
still a bug with a095102, which seems to miss the top right disc when using the default settings

Reply:
Disc?

Reply:
I commented on the wrong PR, oops, I'll move it to the right place.  Resolving.